### PR TITLE
multi refactor: implement pie chart generation vs capacity

### PR DESCRIPTION
**Pull request by BainanXia**
[Pull Request doc](https://breakthrough-energy.github.io/docs/user/git_guide.html#d-pull-request)

### Purpose
Refactor pie chart plot to remove the dependency of `analyze_pg` and other functions/constants in `multi` folder.

### What the code is doing
Implement pie chart plot for generation vs capacity of a list of user specified areas in a list of user specified/customized scenarios

### Testing
A demo notebook is added to illustrate the usage.

### Where to look
```
postreise/plot/plot_pie_generation_vs_capacity.py
postreise/plot/demo/pie_generation_vs_capacity.ipynb
```

### Time estimate
15 min. This one is easier as expected given the major refactor carried out in the previous bar chart PR see [215](https://github.com/Breakthrough-Energy/PostREISE/pull/215)

### CODE REVIEW

File: postreise/plot/plot_pie_generation_vs_capacity.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    :param tuple time_range: [start_timestamp, end_timestamp] where each time stamp
+        is pandas.Timestamp/numpy.datetime64/datetime.datetime. If None, the entire
+        time range is used for the given scenario.
+    :param str time_zone: new time zone, defaults to None, which uses UTC.
+    :param list custom_data: list of dictionaries with each element being
+        hand-generated data as returned by
+        :func:`postreise.plot_bar_generation_vs_capacity.make_gen_cap_custom_data`,
+        defaults to None.
+    :param dict resource_labels: a dictionary with keys being resource types and values
+        being labels, which uses to customize resource labels for selected resource
```
Comment (rouille) [inline postreise/plot/plot_pie_generation_vs_capacity.py:44]: which uses --> which is used

File: postreise/plot/plot_pie_generation_vs_capacity.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        types to show in the plots. Defaults to None, in which case a default set of
+        labels is used.
+    :param dict resource_colors: a dictionary with keys being resource types and values
+        being colors, which uses to customize resource colors for selected resource
```
Comment (rouille) [inline postreise/plot/plot_pie_generation_vs_capacity.py:48]: same here

File: postreise/plot/plot_pie_generation_vs_capacity.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        types to show in the plots. Defaults to None, in which case a default set of
+        colors is used.
+    :param float min_percentage: roll up small pie pieces into a single category,
+        resources with percentage less than the set value will be pooled together,
+        defaults to 0.
+    :raises ValueError:
+        if area_types are provided in a different length of areas and/or
+        if scenario_names are provided in a different length of scenario_ids and/or
+        if less than two scenario_ids and/or custom_data in total is provided and/or
```
Comment (rouille) [inline postreise/plot/plot_pie_generation_vs_capacity.py:57]: extra and/or

File: postreise/plot/plot_pie_generation_vs_capacity.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    :raises TypeError:
+        if resource_labels are provided but not in a dictionary format and/or
+        if resource_colors are provided but not in a dictionary format and/or
```
Comment (rouille) [inline postreise/plot/plot_pie_generation_vs_capacity.py:60]: extra and/or

File: postreise/plot/plot_pie_generation_vs_capacity.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        labels is used.
+    :param dict resource_colors: a dictionary with keys being resource types and values
+        being colors, which uses to customize resource colors for selected resource
+        types to show in the plots. Defaults to None, in which case a default set of
+        colors is used.
+    :param float min_percentage: roll up small pie pieces into a single category,
+        resources with percentage less than the set value will be pooled together,
+        defaults to 0.
+    :raises ValueError:
+        if area_types are provided in a different length of areas and/or
```
Comment (rouille) [inline postreise/plot/plot_pie_generation_vs_capacity.py:55]: if length of area_types and areas are different and/or

File: postreise/plot/plot_pie_generation_vs_capacity.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        if scenario_names are provided in a different length of scenario_ids and/or
```
Comment (rouille) [inline postreise/plot/plot_pie_generation_vs_capacity.py:56]: if length of scenario_names and scenario_ids is different and/or

File: postreise/plot/plot_pie_generation_vs_capacity.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    """
+    if isinstance(areas, str):
+        areas = [areas]
+    if isinstance(area_types, str):
+        area_types = [area_types]
+    if not area_types:
+        area_types = [None] * len(areas)
+    if len(areas) != len(area_types):
+        raise ValueError(
+            "ERROR: if area_types are provided, it should have the same number of entries with areas."
```
Comment (rouille) [inline postreise/plot/plot_pie_generation_vs_capacity.py:75]: ERROR: if area_types are provided, it must have the same number of entries than areas.

File: postreise/plot/plot_pie_generation_vs_capacity.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    if not resource_colors:
+        resource_colors = dict()
+    if not isinstance(resource_colors, dict):
+        raise TypeError("ERROR: resource_colors should be a dictionary")
+
+    type2label.update(resource_labels)
+    type2color.update(resource_colors)
+
+    grid = Grid(["USA"])
+    id2loadzone = grid.id2zone
```
Comment (rouille) [inline postreise/plot/plot_pie_generation_vs_capacity.py:107]: we will need to refactor that

File: postreise/plot/plot_pie_generation_vs_capacity.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+            ),
+            "cap": sum_capacity_by_type_zone(scenario).rename(columns=id2loadzone),
+        }
+        scenario_data[sid] = {
+            "name": scenario_names[i] if scenario_names else scenario.info["name"],
+            "gen": {"label": "Generation", "unit": "TWh", "data": {}},
+            "cap": {"label": "Capacity", "unit": "GW", "data": {}},
+        }
+    for area, area_type in zip(areas, area_types):
+        loadzone_set = area_to_loadzone(grid, area, area_type)
```
Comment (rouille) [inline postreise/plot/plot_pie_generation_vs_capacity.py:124]: we will need to refactor that too

File: postreise/plot/plot_pie_generation_vs_capacity.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+        labels.append(small_category_label)
+
+    return resource_data, labels
+
+
+def _construct_pie_visuals(zone, ax_data_list):
+    """Plot formatted data.
+
+    :param str zone: the zone name
+    :param list ax_data_list: a list of labels, values, and colors for each axis of the
```
Comment (rouille) [inline postreise/plot/plot_pie_generation_vs_capacity.py:210]: We could mention that the element of the list are dictionaries with keys being labels, values and colors

**Comment by rouille** [comment]
Should we have some tests to make sure that the errors are correctly raised and we did not forget any edge cases?

**Comment by rouille** [comment]
Notebook runs as expected. I would add a main title to mirror what we do in the other ones.

File: postreise/plot/plot_pie_generation_vs_capacity.py (hunk 1 above)
Comment (BainanXia) [reply postreise/plot/plot_pie_generation_vs_capacity.py]: Done.

File: postreise/plot/plot_pie_generation_vs_capacity.py (hunk 2 above)
Comment (BainanXia) [reply postreise/plot/plot_pie_generation_vs_capacity.py]: Done.

File: postreise/plot/plot_pie_generation_vs_capacity.py (hunk 3 above)
Comment (BainanXia) [reply postreise/plot/plot_pie_generation_vs_capacity.py]: Done.

File: postreise/plot/plot_pie_generation_vs_capacity.py (hunk 4 above)
Comment (BainanXia) [reply postreise/plot/plot_pie_generation_vs_capacity.py]: Done.

File: postreise/plot/plot_pie_generation_vs_capacity.py (hunk 5 above)
Comment (BainanXia) [reply postreise/plot/plot_pie_generation_vs_capacity.py]: Done.

File: postreise/plot/plot_pie_generation_vs_capacity.py (hunk 6 above)
Comment (BainanXia) [reply postreise/plot/plot_pie_generation_vs_capacity.py]: Done.

File: postreise/plot/plot_pie_generation_vs_capacity.py (hunk 7 above)
Comment (BainanXia) [reply postreise/plot/plot_pie_generation_vs_capacity.py]: Done.

File: postreise/plot/plot_pie_generation_vs_capacity.py (hunk 10 above)
Comment (BainanXia) [reply postreise/plot/plot_pie_generation_vs_capacity.py]: Done.

**Comment by BainanXia** [comment]
> Should we have some tests to make sure that the errors are correctly raised and we did not forget any edge cases?

Tests added.

**Comment by BainanXia** [comment]
> Notebook runs as expected. I would add a main title to mirror what we do in the other ones.

A markdown cell in the notebook is added which serves as a main title.

**Comment by rouille** [comment]
> > Notebook runs as expected. I would add a main title to mirror what we do in the other ones.
> 
> A markdown cell in the notebook is added which serves as a main title.

Pie Pot --> Pie Plot

**Comment by BainanXia** [comment]
> > > Notebook runs as expected. I would add a main title to mirror what we do in the other ones.
> > 
> > 
> > A markdown cell in the notebook is added which serves as a main title.
> 
> Pie Pot --> Pie Plot

Good catch! Fixed.

**Review by rouille — APPROVED** [review]
Looks good
